### ISSUE TITLE
[BUG - Segment Replication + Remote Store] - Support segment replication for system indices

### ISSUE THREAD

**Issue by mch2** [opensearch-project/OpenSearch#8182]
**Describe the bug**
Segment Replication is currently not supported on system indices.  This decision was made for segment replication after failing tests with CCR plugin (issues linked below).

Related issues:
https://github.com/opensearch-project/OpenSearch/issues/6602#issuecomment-1486784566
https://github.com/opensearch-project/OpenSearch/issues/3823
https://github.com/opensearch-project/OpenSearch/issues/8158

**To Reproduce**
1. Update opensearch.yml to set cluster replication strategy as SEGMENT. `cluster.indices.replication.strategy: SEGMENT`
2. create a cluster with security or ccr enabled.
3. Fetch default setting values, replication strategy will be DOCUMENT.

**Expected behavior**
Segment Replication should be a supported strategy for system indices.

**Additional context**
This limitation also prevents system indices from being backed up with remote storage.

**Comment by sachinpkale** [comment]
Supporting remote store for system indices involves creating the repository before we create the index. This means we also need to get repository details as part of cluster settings (today, we accept only repository name). I will create a separate issue for the same.

**Comment by sachinpkale** [comment]
cc: @psychbot

**Comment by sachinpkale** [comment]
Issue specific to remote store: https://github.com/opensearch-project/OpenSearch/issues/8188

**Comment by scrawfor99** [comment]
HI @mch2 and @sachinpkale, looking at this, it seems like we will want to engage from the security side of things. Currently, we restrict modification of the security system index. The concern is escalation of privileges and some other issues that can arise such as desync between copies. 

Can you provide some more details on how you plan to interact with the system indices? Is there an issue outlining the flow you are working on? 

Tagging @peternied for some additional context on concerns associated with this.

**Comment by peternied** [comment]
@scrawfor99 I don't think there is an issue with the security plugin itself - its how plugins can interact with system indices that is breaking down with seg-rep enabled.

@mch2 @psychbot @sachinpkale Please let the security team know if there seems to be a 'governance change' related to these indices, which doesn't seem to be the case and I'd guess would break backward compatibility

**Comment by mch2** [comment]
@scrawfor99 @peternied Thanks for taking a look here!

In short there will be no modification to the content of the system index, only the replication strategy that is used behind the scenes to copy it out to replicas.  Segment Replication is now configurable at a cluster level so we would like any index including system indices to use segrep if enabled.

However there are perf implications here that need to be called out.  SegRep will on average increase the amount of time that replica shards are inconsistent from their primaries due to the time required to copy out segments (time dependent on cluster config/hardware etc). Are there are cases where stronger read/write consistency is desired on the system index?  If this is the case plugins would need to [prioritize primaries ](https://github.com/opensearch-project/OpenSearch/pull/7375) for searches.

There is no strong r/w guarantee today with docrep, unless you use _bulk API with wait_until preference, which won't ack the request until each shard has refreshed on those documents.  This option is not supported with segrep.  Looking through security I don't think this is used.

**Comment by peternied** [comment]
> System indices are generally small and should copy out quickly, so I wouldn't expect this to be of major concern. Are there are cases where stronger read/write consistency is desired on the system index? 

@scrawfor99 What do you think about creating an issue to contemplate how we configure segment replication for the security index?  Building up the architecture documentation of security index <-> in memory security configuration might be a good companion for that issue.

**Comment by scrawfor99** [comment]
> 

Hi @peternied, will do! Thanks for taking a look at this. You have a better understanding of the overall picture involved with these types of changes so I dragged you in. I will make an issue over in the Security repo and link it back here.

**Comment by cwperks** [comment]
@mch2 When there is a change to the security configuration, the security plugin will run an transport action on every node of the cluster to instruct it to read the entire security index and refresh its local cache. See: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/action/configupdate/TransportConfigUpdateAction.java#L126-L131

Tracing through the code this will ultimately do a `MultiGetRequest` to get the different `CTypes` (or configuration types meaning `internalusers`, `roles`, etc.). See https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/configuration/ConfigurationLoaderSecurity7.java#L210-L264

Would there be any risk of the `MultiGetRequest` getting stale data?

**Comment by mch2** [comment]
Thanks @cwperks, yes using prefer primary or primary can be used.  If security is using RefreshPolicy.IMMEDIATE, this will only start the refresh & replication process after a write, so its possible there could be a stale read.

This issue is currently blocked by https://github.com/opensearch-project/OpenSearch/issues/8193.

Steps here.
1. https://github.com/opensearch-project/OpenSearch/issues/8193 - Fix bug with remote store/segrep + system indices.
2. merge fix for this issue to remove override of sys indices to docrep in core.
3. Alert plugin teams of the impact of this change, namely that RefreshPolicy.IMMEDIATE will only refresh primary shards & start replication (eventual consistency) and that RefreshPolicy.WAIT_UNTIL is not supported with segrep.  Instead, plugins will need to prioritize primary shards if r/w consistency guarantee is desired. - https://github.com/opensearch-project/OpenSearch/issues/8211.

**Comment by dbwiddis** [comment]
Joining this conversation coming from Job Scheduler repo and I presume I have similar concerns/questions as @peternied and @cwperks discussed with security plugin.  JS uses a system index for locking, which I do not think would be able to use the `IMMEDIATE` option; but JS locking is used in some performance-sensitive tasks so the impact called out by @mch2 [here](https://github.com/opensearch-project/OpenSearch/issues/8182#issuecomment-1602879714) is a potential concern.

Is there any way to keep the current (2.8) situation as a configuration option (allow users to not use SEGREP for system indices) or are we forced to have these perf impacts if the user is using SEGREP at all?

**Comment by cwperks** [comment]
Thank you @mch2 . The security plugin is using [RefreshPolicy.IMMEDIATE after a change to the security index](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java#L361-L385).

I have opened a [PR on the security repo](https://github.com/opensearch-project/security/pull/2903) to add `_primary` preference for the mget request that security performs on a change to the security index.

**Comment by anasalkouz** [comment]
Closing this, since we are tracking the effort here: https://github.com/opensearch-project/OpenSearch/issues/8193
